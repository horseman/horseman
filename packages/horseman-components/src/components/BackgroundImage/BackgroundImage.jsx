import PropTypes from "prop-types";
import styled from "styled-components";

import { imageType } from "../../types";

const BackgroundImage = styled.div`
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;

  background-image: url(${({ src, srcset, width }) => {
    let url = "";
    if (!width) {
      url = src;
    } else {
      // Sort the srcset by width from smallest to largest
      const sortedSrcset = srcset.slice().sort((a, b) => a.width > b.width);

      // Find the smallest based on on the width
      const filteredSource = sortedSrcset.filter(set => set.width > width);

      // Get the best optimized version of the image to use based on width.
      url =
        filteredSource.length > 0
          ? filteredSource[0].src
          : sortedSrcset.pop().src;
    }

    return url;
  }});
`;

BackgroundImage.defaultProps = {
  srcset: [],
};

BackgroundImage.propTypes = {
  ...imageType,

  /**
   * Will display the background image within srcset that is optimized for the
   * pixel width given here.
   */
  width: PropTypes.number,
};

/**
 * @component
 */
export default BackgroundImage;
