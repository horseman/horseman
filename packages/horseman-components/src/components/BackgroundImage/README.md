A BackgroundImage spans the entire width of its container and applies a
background image. By default the background image has no height, it is intended
to be extended.

### Example
```
const image = {
  src: "http://placehold.it/400x400,
}

<div>
  <BackgroundImage
    bgImage={image}
    >
    <div style={{height: '100px', color: 'white'}}>
    Content for the background image
    </div>
  </BackgroundImage>
</div>
```
