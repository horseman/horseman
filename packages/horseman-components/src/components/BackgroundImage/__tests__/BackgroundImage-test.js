import React from "react";
import { shallow } from "enzyme";

import BackgroundImage from "../";

describe("BackgroundImage", () => {
  const image = {
    src: "default",
    srcset: [
      {
        src: "default",
        width: 100,
      },
      {
        src: "larger",
        width: 200,
      },
      {
        src: "largest",
        width: 300,
      },
    ],
  };
  test("renders correctly", () => {
    const wrapper = shallow(
      <BackgroundImage bgImage={image}>foo</BackgroundImage>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  describe("renders correct src with different srcsets", () => {
    image.srcset.forEach(srcset => {
      test(`test ${srcset.src}`, () => {
        const wrapper = shallow(
          <BackgroundImage bgImage={image} width={srcset.width - 1}>
            foo
          </BackgroundImage>,
        );
        expect(wrapper).toMatchSnapshot();
      });
    });
    test("uses the largest image when width is larger than all available srcsets", () => {
      const wrapper = shallow(
        <BackgroundImage bgImage={image} width={500}>
          foo
        </BackgroundImage>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
