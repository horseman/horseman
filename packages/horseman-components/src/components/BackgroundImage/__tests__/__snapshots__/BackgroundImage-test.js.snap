// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage renders correct src with different srcsets test default 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(default);
}

<div
  className="c0"
  src="default"
  width={99}
>
  foo
</div>
`;

exports[`BackgroundImage renders correct src with different srcsets test larger 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(larger);
}

<div
  className="c0"
  src="default"
  width={199}
>
  foo
</div>
`;

exports[`BackgroundImage renders correct src with different srcsets test largest 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(largest);
}

<div
  className="c0"
  src="default"
  width={299}
>
  foo
</div>
`;

exports[`BackgroundImage renders correct src with different srcsets uses the largest image when width is larger than all available srcsets 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(largest);
}

<div
  className="c0"
  src="default"
  width={500}
>
  foo
</div>
`;

exports[`BackgroundImage renders correctly 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(default);
}

<div
  className="c0"
  src="default"
>
  foo
</div>
`;

exports[`BackgroundImage renders correctly without a srcset passed in 1`] = `
.c0 {
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  background-image: url(default);
}

<div
  className="c0"
  src="default"
  width={10}
>
  foo
</div>
`;
